demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -223,23 +498,63 @@ static int _discoverGreengrassCore( const IotNetworkInterface_t * pNetworkInterf
         == pdPASS )
     {
         IotLogInfo( "Greengrass device discovered." );
-        mqttStatus = _mqttConnect( &xHostAddressData, pNetworkInterface, &mqttConnection );
 
-        if( mqttStatus == IOT_MQTT_SUCCESS )
+        /* Set the entry time of the demo application. This entry time will be used
+         * to calculate relative time elapsed in the execution of the demo application,
+         * by the timer utility function that is provided to the MQTT library.
+         */
+        ulGlobalEntryTimeMs = prvGetTimeMs();
+        xNetworkContext.pParams = &secureSocketsTransportParams;
+
+        /****************************** Connect. ******************************/
+
+        /* Attempt to establish TLS session with MQTT broker. If connection fails,
+         * retry after a timeout. Timeout value will be exponentially increased until
+         * the maximum number of attempts are reached or the maximum timeout value is reached.
+         * The function returns a failure status if the TLS over TCP connection cannot be established
+         * to the broker after the configured number of attempts. */
+        xDemoStatus = prvConnectToServerWithBackoffRetries( &xHostAddressData, &xNetworkContext, &xMQTTContext );
+
+        if( xDemoStatus == pdPASS )
+        {
+            xIsConnectionEstablished = pdTRUE;
+        }
+        else
+        {
+            IotLogError( "Could not establish TLS session with MQTT broker." );
+            status = EXIT_FAILURE;
+        }
+
+        if( xDemoStatus == pdPASS )
         {
-            _sendMessageToGGC( mqttConnection );
+            _sendMessageToGGC( &xMQTTContext );
 
             IotLogInfo( "Disconnecting from broker." );
 
             /* Call MQTT v2's DISCONNECT function. */
-            IotMqtt_Disconnect( mqttConnection, 0 );
-            mqttConnection = IOT_MQTT_CONNECTION_INITIALIZER;
+            MQTT_Disconnect( &xMQTTContext );
+
+            /* We will always close the network connection, even if an error may have occurred during
+             * demo execution, to clean up the system resources that it may have consumed. */
+            if( xIsConnectionEstablished == pdTRUE )
+            {
+                /* Close the network connection.  */
+                xNetworkStatus = SecureSocketsTransport_Disconnect( &xNetworkContext );
+
+                if( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS )
+                {
+                    xDemoStatus = pdFAIL;
+                    status = EXIT_FAILURE;
```

Comment:
Is there a need to have both a status and xDemoStatus? Doesn't seem like both are needed.